test/e2e_node/criproxy/image.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package criproxy
+
+import (
+	"context"
+
+	internalapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+// ImageServiceProxy wraps the ImageManagerService and injects exceptions (such as errors or delays)
+type ImageServiceProxy struct {
+	injectors           []func(string) error
+	imageManagerService internalapi.ImageManagerService
+}
+
+var _ internalapi.ImageManagerService = (*ImageServiceProxy)(nil)
```

Comment:
Thanks, I think this approach is more common than embedding the interface. If you want to enforce the implementation of the interface this way, it seems like you could simplify it as follows:


<details>
<summary>Code changes</summary>


```diff
diff --git a/test/e2e_node/criproxy/image.go b/test/e2e_node/criproxy/image.go
index 0d24efac6f1..a7e6c9f9481 100644
--- a/test/e2e_node/criproxy/image.go
+++ b/test/e2e_node/criproxy/image.go
@@ -25,17 +25,14 @@ import (

 // ImageServiceProxy wraps the ImageManagerService and injects exceptions (such as errors or delays)
 type ImageServiceProxy struct {
-       injectors           []func(string) error
-       imageManagerService internalapi.ImageManagerService
+       injectors []func(string) error
 }

 var _ internalapi.ImageManagerService = (*ImageServiceProxy)(nil)
 var _ CriProxy = (*ImageServiceProxy)(nil)

-func NewImageServiceProxy(client internalapi.ImageManagerService) internalapi.ImageManagerService {
-       return &ImageServiceProxy{
-               imageManagerService: client,
-       }
+func NewImageServiceProxy() internalapi.ImageManagerService {
+       return &ImageServiceProxy{}
 }

 func (i *ImageServiceProxy) AddInjector(injector func(string) error) {
@@ -46,35 +43,35 @@ func (i *ImageServiceProxy) ListImages(ctx context.Context, filter *runtimeapi.I
        if err := i.runInjector("ListImages"); err != nil {
                return nil, err
        }
-       return i.imageManagerService.ListImages(ctx, filter)
+       return i.ListImages(ctx, filter)
 }

 func (i *ImageServiceProxy) ImageStatus(ctx context.Context, image *runtimeapi.ImageSpec, verbose bool) (*runtimeapi.ImageStatusResponse, error) {
        if err := i.runInjector("ImageStatus"); err != nil {
                return nil, err
        }
-       return i.imageManagerService.ImageStatus(ctx, image, verbose)
+       return i.ImageStatus(ctx, image, verbose)
 }

 func (i *ImageServiceProxy) PullImage(ctx context.Context, image *runtimeapi.ImageSpec, auth *runtimeapi.AuthConfig, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error) {
        if err := i.runInjector("PullImage"); err != nil {
                return "", err
        }
-       return i.imageManagerService.PullImage(ctx, image, auth, podSandboxConfig)
+       return i.PullImage(ctx, image, auth, podSandboxConfig)
 }

 func (i *ImageServiceProxy) RemoveImage(ctx context.Context, image *runtimeapi.ImageSpec) error {
        if err := i.runInjector("RemoveImage"); err != nil {
                return err
        }
-       return i.imageManagerService.RemoveImage(ctx, image)
+       return i.RemoveImage(ctx, image)
 }

 func (i *ImageServiceProxy) ImageFsInfo(ctx context.Context) (*runtimeapi.ImageFsInfoResponse, error) {
        if err := i.runInjector("ImageFsInfo"); err != nil {
                return nil, err
        }
-       return i.imageManagerService.ImageFsInfo(ctx)
+       return i.ImageFsInfo(ctx)
 }

 func (i *ImageServiceProxy) runInjector(apiName string) error {
diff --git a/test/e2e_node/criproxy/runtime.go b/test/e2e_node/criproxy/runtime.go
index a3cbdc1cc41..fb223781b1b 100644
--- a/test/e2e_node/criproxy/runtime.go
+++ b/test/e2e_node/criproxy/runtime.go
@@ -26,17 +26,14 @@ import (

 // RuntimeServiceProxy wraps the RuntimeService and injects exceptions (such as errors or delays)
 type RuntimeServiceProxy struct {
-       injectors      []func(string) error
-       runtimeService internalapi.RuntimeService
+       injectors []func(string) error
 }

 var _ internalapi.RuntimeService = (*RuntimeServiceProxy)(nil)
 var _ CriProxy = (*RuntimeServiceProxy)(nil)

-func NewRuntimeServiceProxy(client internalapi.RuntimeService) internalapi.RuntimeService {
-       return &RuntimeServiceProxy{
-               runtimeService: client,
-       }
+func NewRuntimeServiceProxy() internalapi.RuntimeService {
+       return &RuntimeServiceProxy{}
 }

 func (r *RuntimeServiceProxy) AddInjector(injector func(string) error) {
@@ -47,202 +44,202 @@ func (r *RuntimeServiceProxy) Version(ctx context.Context, apiVersion string) (*
        if err := r.runInjector("Version"); err != nil {
                return nil, err
        }
-       return r.runtimeService.Version(ctx, apiVersion)
+       return r.Version(ctx, apiVersion)
 }

 func (r *RuntimeServiceProxy) RunPodSandbox(ctx context.Context, config *runtimeapi.PodSandboxConfig, runtimeHandler string) (string, error) {
        if err := r.runInjector("RunPodSandbox"); err != nil {
                return "", err
        }
-       return r.runtimeService.RunPodSandbox(ctx, config, runtimeHandler)
+       return r.RunPodSandbox(ctx, config, runtimeHandler)
 }

 func (r *RuntimeServiceProxy) StopPodSandbox(ctx context.Context, podSandBoxID string) (err error) {
        if err := r.runInjector("StopPodSandbox"); err != nil {
                return err
        }
-       return r.runtimeService.StopPodSandbox(ctx, podSandBoxID)
+       return r.StopPodSandbox(ctx, podSandBoxID)
 }

 func (r *RuntimeServiceProxy) RemovePodSandbox(ctx context.Context, podSandboxID string) error {
        if err := r.runInjector("RemovePodSandbox"); err != nil {
                return err
        }
-       return r.runtimeService.RemovePodSandbox(ctx, podSandboxID)
+       return r.RemovePodSandbox(ctx, podSandboxID)
 }

 func (r *RuntimeServiceProxy) PodSandboxStatus(ctx context.Context, podSandboxID string, verbose bool) (*runtimeapi.PodSandboxStatusResponse, error) {
        if err := r.runInjector("PodSandboxStatus"); err != nil {
                return nil, err
        }
-       return r.runtimeService.PodSandboxStatus(ctx, podSandboxID, verbose)
+       return r.PodSandboxStatus(ctx, podSandboxID, verbose)
 }

 func (r *RuntimeServiceProxy) ListPodSandbox(ctx context.Context, filter *runtimeapi.PodSandboxFilter) ([]*runtimeapi.PodSandbox, error) {
        if err := r.runInjector("ListPodSandbox"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ListPodSandbox(ctx, filter)
+       return r.ListPodSandbox(ctx, filter)
 }

 func (r *RuntimeServiceProxy) PortForward(ctx context.Context, request *runtimeapi.PortForwardRequest) (*runtimeapi.PortForwardResponse, error) {
        if err := r.runInjector("PortForward"); err != nil {
                return nil, err
        }
-       return r.runtimeService.PortForward(ctx, request)
+       return r.PortForward(ctx, request)
 }

 func (r *RuntimeServiceProxy) CreateContainer(ctx context.Context, podSandboxID string, config *runtimeapi.ContainerConfig, sandboxConfig *runtimeapi.PodSandboxConfig) (string, error) {
        if err := r.runInjector("CreateContainer"); err != nil {
                return "", err
        }
-       return r.runtimeService.CreateContainer(ctx, podSandboxID, config, sandboxConfig)
+       return r.CreateContainer(ctx, podSandboxID, config, sandboxConfig)
 }

 func (r *RuntimeServiceProxy) StartContainer(ctx context.Context, containerID string) error {
        if err := r.runInjector("StartContainer"); err != nil {
                return err
        }
-       return r.runtimeService.StartContainer(ctx, containerID)
+       return r.StartContainer(ctx, containerID)
 }

 func (r *RuntimeServiceProxy) StopContainer(ctx context.Context, containerID string, timeout int64) error {
        if err := r.runInjector("StopContainer"); err != nil {
                return err
        }
-       return r.runtimeService.StopContainer(ctx, containerID, timeout)
+       return r.StopContainer(ctx, containerID, timeout)
 }

 func (r *RuntimeServiceProxy) RemoveContainer(ctx context.Context, containerID string) error {
        if err := r.runInjector("RemoveContainer"); err != nil {
                return err
        }
-       return r.runtimeService.RemoveContainer(ctx, containerID)
+       return r.RemoveContainer(ctx, containerID)
 }

 func (r *RuntimeServiceProxy) ListContainers(ctx context.Context, filter *runtimeapi.ContainerFilter) ([]*runtimeapi.Container, error) {
        if err := r.runInjector("ListContainers"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ListContainers(ctx, filter)
+       return r.ListContainers(ctx, filter)
 }

 func (r *RuntimeServiceProxy) ContainerStatus(ctx context.Context, containerID string, verbose bool) (*runtimeapi.ContainerStatusResponse, error) {
        if err := r.runInjector("ContainerStatus"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ContainerStatus(ctx, containerID, verbose)
+       return r.ContainerStatus(ctx, containerID, verbose)
 }

 func (r *RuntimeServiceProxy) UpdateContainerResources(ctx context.Context, containerID string, resources *runtimeapi.ContainerResources) error {
        if err := r.runInjector("UpdateContainerResources"); err != nil {
                return err
        }
-       return r.runtimeService.UpdateContainerResources(ctx, containerID, resources)
+       return r.UpdateContainerResources(ctx, containerID, resources)
 }

 func (r *RuntimeServiceProxy) ExecSync(ctx context.Context, containerID string, cmd []string, timeout time.Duration) (stdout []byte, stderr []byte, err error) {
        if err := r.runInjector("ExecSync"); err != nil {
                return nil, nil, err
        }
-       return r.runtimeService.ExecSync(ctx, containerID, cmd, timeout)
+       return r.ExecSync(ctx, containerID, cmd, timeout)
 }

 func (r *RuntimeServiceProxy) Exec(ctx context.Context, request *runtimeapi.ExecRequest) (*runtimeapi.ExecResponse, error) {
        if err := r.runInjector("Exec"); err != nil {
                return nil, err
        }
-       return r.runtimeService.Exec(ctx, request)
+       return r.Exec(ctx, request)
 }

 func (r *RuntimeServiceProxy) Attach(ctx context.Context, req *runtimeapi.AttachRequest) (*runtimeapi.AttachResponse, error) {
        if err := r.runInjector("Attach"); err != nil {
                return nil, err
        }
-       return r.runtimeService.Attach(ctx, req)
+       return r.Attach(ctx, req)
 }

 func (r *RuntimeServiceProxy) ReopenContainerLog(ctx context.Context, containerID string) error {
        if err := r.runInjector("ReopenContainerLog"); err != nil {
                return err
        }
-       return r.runtimeService.ReopenContainerLog(ctx, containerID)
+       return r.ReopenContainerLog(ctx, containerID)
 }

 func (r *RuntimeServiceProxy) CheckpointContainer(ctx context.Context, options *runtimeapi.CheckpointContainerRequest) error {
        if err := r.runInjector("CheckpointContainer"); err != nil {
                return err
        }
-       return r.runtimeService.CheckpointContainer(ctx, options)
+       return r.CheckpointContainer(ctx, options)
 }

 func (r *RuntimeServiceProxy) GetContainerEvents(ctx context.Context, containerEventsCh chan *runtimeapi.ContainerEventResponse, connectionEstablishedCallback func(runtimeapi.RuntimeService_GetContainerEventsClient)) error {
        if err := r.runInjector("GetContainerEvents"); err != nil {
                return err
        }
-       return r.runtimeService.GetContainerEvents(ctx, containerEventsCh, connectionEstablishedCallback)
+       return r.GetContainerEvents(ctx, containerEventsCh, connectionEstablishedCallback)
 }

 func (r *RuntimeServiceProxy) ContainerStats(ctx context.Context, containerID string) (*runtimeapi.ContainerStats, error) {
        if err := r.runInjector("ContainerStats"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ContainerStats(ctx, containerID)
+       return r.ContainerStats(ctx, containerID)
 }
 func (r *RuntimeServiceProxy) ListContainerStats(ctx context.Context, filter *runtimeapi.ContainerStatsFilter) ([]*runtimeapi.ContainerStats, error) {
        if err := r.runInjector("ListContainerStats"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ListContainerStats(ctx, filter)
+       return r.ListContainerStats(ctx, filter)
 }

 func (r *RuntimeServiceProxy) PodSandboxStats(ctx context.Context, podSandboxID string) (*runtimeapi.PodSandboxStats, error) {
        if err := r.runInjector("PodSandboxStats"); err != nil {
                return nil, err
        }
-       return r.runtimeService.PodSandboxStats(ctx, podSandboxID)
+       return r.PodSandboxStats(ctx, podSandboxID)
 }

 func (r *RuntimeServiceProxy) ListPodSandboxStats(ctx context.Context, filter *runtimeapi.PodSandboxStatsFilter) ([]*runtimeapi.PodSandboxStats, error) {
        if err := r.runInjector("ListPodSandboxStats"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ListPodSandboxStats(ctx, filter)
+       return r.ListPodSandboxStats(ctx, filter)
 }

 func (r *RuntimeServiceProxy) ListMetricDescriptors(ctx context.Context) ([]*runtimeapi.MetricDescriptor, error) {
        if err := r.runInjector("ListMetricDescriptors"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ListMetricDescriptors(ctx)
+       return r.ListMetricDescriptors(ctx)
 }

 func (r *RuntimeServiceProxy) ListPodSandboxMetrics(ctx context.Context) ([]*runtimeapi.PodSandboxMetrics, error) {
        if err := r.runInjector("ListPodSandboxMetrics"); err != nil {
                return nil, err
        }
-       return r.runtimeService.ListPodSandboxMetrics(ctx)
+       return r.ListPodSandboxMetrics(ctx)
 }

 func (r *RuntimeServiceProxy) UpdateRuntimeConfig(ctx context.Context, runtimeConfig *runtimeapi.RuntimeConfig) error {
        if err := r.runInjector("UpdateRuntimeConfig"); err != nil {
                return err
        }
-       return r.runtimeService.UpdateRuntimeConfig(ctx, runtimeConfig)
+       return r.UpdateRuntimeConfig(ctx, runtimeConfig)
 }

 func (r *RuntimeServiceProxy) Status(ctx context.Context, verbose bool) (*runtimeapi.StatusResponse, error) {
        if err := r.runInjector("Status"); err != nil {
                return nil, err
        }
-       return r.runtimeService.Status(ctx, verbose)
+       return r.Status(ctx, verbose)
 }

 func (r *RuntimeServiceProxy) RuntimeConfig(ctx context.Context) (*runtimeapi.RuntimeConfigResponse, error) {
        if err := r.runInjector("RuntimeConfig"); err != nil {
                return nil, err
        }
-       return r.runtimeService.RuntimeConfig(ctx)
+       return r.RuntimeConfig(ctx)
 }

 func (r *RuntimeServiceProxy) runInjector(apiName string) error {
diff --git a/test/e2e_node/util.go b/test/e2e_node/util.go
index bad7c3bd238..1eed87017f8 100644
--- a/test/e2e_node/util.go
+++ b/test/e2e_node/util.go
@@ -337,8 +337,8 @@ func getCRIClient() (internalapi.RuntimeService, internalapi.ImageManagerService
                return nil, nil, err
        }
        if framework.TestContext.CriProxyEnabled {
-               runtimeServiceProxy := criproxy.NewRuntimeServiceProxy(r)
-               imageServiceProxy := criproxy.NewImageServiceProxy(i)
+               runtimeServiceProxy := criproxy.NewRuntimeServiceProxy()
+               imageServiceProxy := criproxy.NewImageServiceProxy()
                return runtimeServiceProxy, imageServiceProxy, nil
        }
        return r, i, nil
```

</details>